test/experimental/features/new_trigger_filters/filter_constructors.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2023 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package new_trigger_filters
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+	"text/template"
+)
+
+type Filter interface {
+	FilterString(indents int) string
+}
+
+type AttributeFilter struct {
+	Type      string
+	Attribute string
+	Value     string
+}
+
+func (f *AttributeFilter) FilterString(indents int) string {
+	return indent(fmt.Sprintf(`- %s:
+	%s: %s`, f.Type, f.Attribute, f.Value), indents)
+}
+
+type CESQLFilter struct {
+	Value string
+}
+
+func (f *CESQLFilter) FilterString(indents int) string {
+	return indent(fmt.Sprintf(" - cesql: %s", f.Value), indents)
+}
+
+type NotFilter struct {
+	Filter Filter
+}
+
+func (f *NotFilter) FilterString(indents int) string {
+	tmpl, err := template.New("not-filter").Parse("- not:\n{{ .Filter.FilterString 1 }}")
+	if err != nil {
+		panic("failed to parse filter string")
+	}
+	var result bytes.Buffer
+	err = tmpl.Execute(&result, *f)
+	if err != nil {
+		panic("failed to execute template to filter")
+	}
+	return indent(result.String(), indents)
+}
+
+type ArrayFilter struct {
+	Type    string
+	Filters []Filter
+}
+
+func (f *ArrayFilter) FilterString(indents int) string {
+	var helpers template.FuncMap = map[string]interface{}{
+		"last": func(index int, len int) bool {
+			return index == len-1
+		},
+	}
+	tmpl, err := template.New("filter").Funcs(helpers).Parse("- {{.Type}}:\n{{$filtersLength := len .Filters}}{{range $idx, $filter := .Filters}}{{$filter.FilterString 1}}{{if not (last $idx $filtersLength)}}\n{{end}}{{end}}")
+	if err != nil {
+		panic("failed to parse filter string")
```

Comment:
Change to `panic(fmt.Sprintf("failed to parse filter string: %v", err))` can reveal more error information